### PR TITLE
use context to derive app env (prod vs staging)

### PR DESCRIPTION
**Pull request by amirbey**

### CODE REVIEW

File: lambdas/helpers/dynamoDBHelper.js [hunk 1]
```diff
@@ -49,8 +54,54 @@ const parseURI = (request) => {
...
+const getDomain = (functionName) => {
+  let domain;
+  const appEnv = getAppEnv(functionName);
+  if appEnv === 'staging' {
+    domain = 'sites-staging.federalist.18f.gov';
+  }
+  if appEnv === 'prod' {
+    domain = 'sites-prod.federalist.18f.gov';
+  }
+  return tableName;
```
Comment (davemcorwin) [inline lambdas/helpers/dynamoDBHelper.js]: should this be `return domain` ?

File: lambdas/helpers/utils.js [hunk 2]
```diff
@@ -0,0 +1,3 @@
+const getHost = request => request.headers.host[0].value;
+
+module.exports = { getHost };
```
Comment (davemcorwin) [inline lambdas/helpers/utils.js:3]: Doesn't look like this is being used anywhere... the same function exists in app.js

File: lambdas/helpers/dynamoDBHelper.js [hunk 3]
```diff
@@ -2,40 +2,79 @@
 const AWS = require('aws-sdk');
 const log = require('./logger');
 
+const testConfig = {
```
Comment (davemcorwin) [inline lambdas/helpers/dynamoDBHelper.js]: I think we can make this even simpler. I don't think the `appEnv` key is used, but we can use it to make one big config object while making sure the env keys match exactly with the env from the function names:
```
const config = {
  prod: {
    domain: 'sites-prod.federalist.18f.gov',
    tableName: 'federalist-proxy-prod',
    siteKey: 'id',
  },
  staging: {
    domain: 'sites-staging.federalist.18f.gov',
    tableName: 'federalist-proxy-staging',
    siteKey: 'id',
  },
  test: {
    domain: 'sites-test.federalist.18f.gov',
    tableName: 'federalist-proxy-test',
    siteKey: 'id',
  },
}
```

Then, when we check the function name against the regular expression, we can capture the value of the environment and use it to look up the config in the object so `getAppConfig` could look like:

```
const getAppConfig = (functionName) => {
  const [, appEnv] = re.exec(functionName);
  return config[appEnv] || {};
};
```

where `re` is just the regular expression:

```
const re = /^us-east-1:federalist-proxy-(prod|staging|test)-(viewer|origin)-(request|response):\d+$/;
```

This also has the advantage of only checking 1 regex.

File: lambdas/helpers/utils.js (hunk 2 above)
Comment (davemcorwin) [reply lambdas/helpers/utils.js:3]: This logic looks to be repeated in line 74 of app.js as well

File: lambdas/helpers/dynamoDBHelper.js [hunk 4]
```diff
@@ -2,40 +2,79 @@
...
+    } else if (/prod/.test(functionName)) {
+      appConfig = prodConfig;
+    } else if (/test/.test(functionName)) {
+      appConfig = testConfig;
+    }
+  }
+  return appConfig;
+};
 
+const getSiteKeyValue = (host, appDomain) => {
```
Comment (davemcorwin) [inline lambdas/helpers/dynamoDBHelper.js]: It looks like this function strips the domain from the end of the host if it is present or returns undefined.

Wondering if:
1. Renaming this function something like `stripDomain` would be more semantic.
2. It should throw an error if the host doesn't end in the domain OR not even check and let it not find an entry in the db. Does it error right now if it looks up a records with a key === `undefined` in Dynamo?

**Review by davemcorwin — COMMENTED** [review]
Thanks @amirbey have some detailed comments below for you to take a look at.

File: lambdas/helpers/utils.js (hunk 2 above)
Comment (amirbey) [reply lambdas/helpers/utils.js:3]: thanks ... i meant to move from app.js to utils.js

File: lambdas/helpers/dynamoDBHelper.js [hunk 5]
```diff
@@ -2,40 +2,71 @@
...
+const getAppConfig = (functionName) => {
+  if (validateProxyFunctionName(functionName)) {
+    const appEnv = /prod|staging|test/.exec(functionName);
+    if (appEnv && appEnv.length) {
+      return appConfig[appEnv[0]];
+    }
+  }
+  throw new Error(`Unable to find appConfig for @function: ${functionName}`);
+};
+
```
Comment (davemcorwin) [inline lambdas/helpers/dynamoDBHelper.js:51]: Why do we need to test the regex twice? Can't we just:
```
const functionRE =  /^us-east-1:federalist-proxy-(prod|staging|test)-(viewer|origin)-(request|response):\d+$/;

const getAppConfig = (functionName) => {
  const match = functionRE.exec(functionName);
  if (match === null) { // or `if (!match)`
    throw new Error(`Unable to find appConfig for @function: ${functionName}`);
  }
  return appConfig[appEnv[1]];
};
```

Comment (amirbey) [reply lambdas/helpers/dynamoDBHelper.js:51]: @davemcorwin 
in my code, the first check is to validate the function and the second gets the appEnv
With your suggestion how are you defining appEnv ?

Comment (davemcorwin) [reply lambdas/helpers/dynamoDBHelper.js:51]: Both are checking the regex using `.exec()`, if the regex doesn't match, it will return null.

**Review by davemcorwin — APPROVED** [review]
💯

**Comment by amirbey** [comment]
👀 @davemcorwin  https://github.com/18F/federalist-proxy-edge/commit/ccce36a7490a4fcd54d742495c2d9b11db256e7f

**Review by davemcorwin — APPROVED** [review]
👍

**Pull request merged**